xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -133,6 +209,162 @@ bool CalculatesExpectation(const HloInstruction* instr) {
          ((actual_r_nelems + r_nelems) * numerical_epsilon);
 }
 
+// Recursively traverses the graph across converts, bitcasts and reshapes
+// starting from instr, and returns the operand of custom_call with index
+// operand_idx. Returns nullptr if the prescribed operand of custom_call cannot
+// be reached from instr.
+HloInstruction* FindOperandRecursive(HloInstruction* instr,
+                                     HloInstruction* custom_call,
+                                     absl::flat_hash_set<int>& visited_instrs,
```

Comment:
visited_instrs should have HloInstruction*, not ints representing the unique IDs of instructions. And similarly for FindAddReduceRecursive.